src/libaktualizr-posix/ipuptanesecondary.cc

```diff
@@ -52,6 +52,40 @@ Uptane::SecondaryInterface::Ptr IpUptaneSecondary::create(const std::string& add
   return std::make_shared<IpUptaneSecondary>(address, port, serial, hw_id, pub_key);
 }
 
+SecondaryInterface::Ptr IpUptaneSecondary::connectAndCheck(const std::string& address, unsigned short port,
+                                                           EcuSerial serial, HardwareIdentifier hw_id,
+                                                           PublicKey pub_key) {
+  // try to connect:
+  // - if it succeeds compare with what we expect
+  // - otherwise, keep using what we know
+  try {
+    auto sec = IpUptaneSecondary::connectAndCreate(address, port);
+    if (sec != nullptr) {
+      auto s = sec->getSerial();
+      if (s != serial) {
+        LOG_ERROR << "Mismatch between secondary serials " << s << " and " << serial;
+        return nullptr;
+      }
+      auto h = sec->getHwId();
+      if (h != hw_id) {
+        LOG_ERROR << "Mismatch between hardware ids " << h << " and " << hw_id;
+        return nullptr;
+      }
+      auto p = sec->getPublicKey();
+      if (p != pub_key) {
+        LOG_WARNING << "Mismatch between public keys " << p.Value() << " and " << pub_key.Value() << " for secondary "
+                    << serial;
```

Comment:
Should we store the new key in this case? Or is that already covered by `SotaUptaneClient::addSecondary`?

Reply:
The storage part is indeed covered in `SotaUptaneClient` but I've added another commit to try to cover more cases.